### PR TITLE
[SPARK-42782][SQL][TESTS] Hive compatibility check for get_json_object

### PR DESCRIPTION
**Pull request by wangyum**
### What changes were proposed in this pull request?

This PR ports the [tests](https://github.com/apache/hive/blob/rel/release-3.1.3/ql/src/test/org/apache/hadoop/hive/ql/udf/TestUDFJson.java) for `get_json_object` from the Apache Hive project.

### Why are the changes needed?

Increase test coverage.

### Does this PR introduce _any_ user-facing change?

No.

### How was this patch tested?

N/A.

### CODE REVIEW

File: sql/core/src/test/scala/org/apache/spark/sql/JsonFunctionsSuite.scala [hunk 1]
```diff
@@ -81,6 +81,93 @@ class JsonFunctionsSuite extends QueryTest with SharedSparkSession {
...
+    runTest("[\"a\",\"b\"]", "$[1]", "b")
+    runTest("[[\"a\",\"b\"]]", "$[0][1]", "b")
+
+    runTest("[1,2,3]", "[0]", null)
+    runTest("[1,2,3]", "$0", null)
+    runTest("[1,2,3]", "0", null)
+    runTest("[1,2,3]", "$.", null)
+
+    runTest("[1,2,3]", "$", "[1,2,3]")
+    runTest("{\"a\":4}", "$", "{\"a\":4}")
```
Comment (wangyum) [inline sql/core/src/test/scala/org/apache/spark/sql/JsonFunctionsSuite.scala:162]: Hive's test: https://github.com/apache/hive/blob/rel/release-3.1.3/ql/src/test/org/apache/hadoop/hive/ql/udf/TestUDFJson.java#L31-L110

File: sql/core/src/test/scala/org/apache/spark/sql/JsonFunctionsSuite.scala [hunk 2]
```diff
@@ -81,6 +81,93 @@ class JsonFunctionsSuite extends QueryTest with SharedSparkSession {
...
+    runTest(json, "$.store.basket[0][*].non_exist_key", null)
+    runTest(json, "$.store.basket[*][*].non_exist_key", null)
+    runTest(json, "$.zip code", "94025")
+    runTest(json, "$.fb:testid", "1234")
+    runTest("{\"a\":\"b\nc\"}", "$.a", "b\nc")
+
+    // Test root array
+    runTest("[1,2,3]", "$[0]", "1")
+    runTest("[1,2,3]", "$.[0]", null) // Not supported
+    runTest("[1,2,3]", "$.[1]", null) // Not supported
```
Comment (wangyum) [inline sql/core/src/test/scala/org/apache/spark/sql/JsonFunctionsSuite.scala:136]: Spark does not support.

File: sql/core/src/test/scala/org/apache/spark/sql/JsonFunctionsSuite.scala [hunk 3]
```diff
@@ -81,6 +81,93 @@ class JsonFunctionsSuite extends QueryTest with SharedSparkSession {
...
+    runTest("[1,2,3]", "$[1]", "2")
+
+    runTest("[1,2,3]", "$[3]", null)
+    runTest("[1,2,3]", "$.[*]", null) // Not supported
```
Comment (wangyum) [inline sql/core/src/test/scala/org/apache/spark/sql/JsonFunctionsSuite.scala:140]: Spark does not support.

File: sql/core/src/test/scala/org/apache/spark/sql/JsonFunctionsSuite.scala [hunk 4]
```diff
@@ -81,6 +81,90 @@ class JsonFunctionsSuite extends QueryTest with SharedSparkSession {
...
+    runTest(json, "$.store.basket[0][*].non_exist_key", null)
+    runTest(json, "$.store.basket[*][*].non_exist_key", null)
+    runTest(json, "$.zip code", "94025")
+    runTest(json, "$.fb:testid", "1234")
+    runTest("{\"a\":\"b\nc\"}", "$.a", "b\nc")
+
+    // Test root array
+    runTest("[1,2,3]", "$[0]", "1")
+    runTest("[1,2,3]", "$.[0]", null) // Not supported
+    runTest("[1,2,3]", "$.[1]", null) // Not supported
```
Comment (wangyum) [inline sql/core/src/test/scala/org/apache/spark/sql/JsonFunctionsSuite.scala:134]: Spark does not support.

File: sql/core/src/test/scala/org/apache/spark/sql/JsonFunctionsSuite.scala [hunk 5]
```diff
@@ -81,6 +81,90 @@ class JsonFunctionsSuite extends QueryTest with SharedSparkSession {
...
+    runTest("[1,2,3]", "$[1]", "2")
+
+    runTest("[1,2,3]", "$[3]", null)
+    runTest("[1,2,3]", "$.[*]", null) // Not supported
```
Comment (wangyum) [inline sql/core/src/test/scala/org/apache/spark/sql/JsonFunctionsSuite.scala:138]: Spark does not support.

File: sql/core/src/test/scala/org/apache/spark/sql/JsonFunctionsSuite.scala [hunk 6]
```diff
@@ -81,6 +81,90 @@ class JsonFunctionsSuite extends QueryTest with SharedSparkSession {
       expected)
   }
 
+  test("SPARK-42782: Port get_json_object's tests from the Apache Hive project") {
```
Comment (wangyum) [inline sql/core/src/test/scala/org/apache/spark/sql/JsonFunctionsSuite.scala:84]: Hive's test: https://github.com/apache/hive/blob/rel/release-3.1.3/ql/src/test/org/apache/hadoop/hive/ql/udf/TestUDFJson.java#L31-L110

Comment (HyukjinKwon) [inline sql/core/src/test/scala/org/apache/spark/sql/JsonFunctionsSuite.scala:84]: Why don't you name this test title like "SPARK-42782: Hive compatibility check for get_json_object"

**Review by HyukjinKwon — APPROVED** [review]

File: sql/core/src/test/scala/org/apache/spark/sql/JsonFunctionsSuite.scala (hunk 6 above)
Comment (wangyum) [reply sql/core/src/test/scala/org/apache/spark/sql/JsonFunctionsSuite.scala:84]: Do you prefer "[SPARK-42782](https://issues.apache.org/jira/browse/SPARK-42782): Hive compatibility check for get_json_object"?

Comment (HyukjinKwon) [reply sql/core/src/test/scala/org/apache/spark/sql/JsonFunctionsSuite.scala:84]: Yeah but I don't mind if you don't prefer that. Will leave it to you.

Comment (wangyum) [reply sql/core/src/test/scala/org/apache/spark/sql/JsonFunctionsSuite.scala]: Renamed it.

**Review by dongjoon-hyun — APPROVED** [review]
+1, LGTM. Thank you for adding this, @wangyum and @HyukjinKwon .
Merged to master for Apache Spark 3.5.

**Pull request closed without merging**